pennylane_ionq/device.py

```diff
@@ -235,34 +380,164 @@ def _submit_job(self):
+        prob_array = np.zeros([2**self.num_wires])
+
+        # histogram values don't always perfectly sum to exactly one
+        histogram_values = self.histogram.values()
+        norm = sum(histogram_values)
+        prob_array[idx] = np.fromiter(histogram_values, float) / norm
 
-            # convert the sparse probs into a probability array
-            self._prob_array = np.zeros([2**self.num_wires])
+        if self._current_circuit_index is not None:
+            self.histogram = None
 
-            # histogram values don't always perfectly sum to exactly one
-            histogram_values = self.histogram.values()
-            norm = sum(histogram_values)
-            self._prob_array[idx] = np.fromiter(histogram_values, float) / norm
+        return prob_array
+
+    def estimate_probability(
+        self, wires=None, shot_range=None, bin_size=None
+    ):
+        """Return the estimated probability of each computational basis state
+        using the generated samples. Overwrites the method in QubitDevice class
+        to accomodate batch submission of circuits. When invoking this method after 
+        submitting circuits using any other method than QubitDevice.execute() 
+        method, current circuit index should be set via set_current_circuit_index() 
+        method to the circuit which is targeted out of one or potentially several 
+        circuits submitted in a batch. This is not very good design but the 
+        alternative would be to create huge amounts of code duplication with respect
+        to the Pennylane base code.
+
+        Args:
+            wires (Iterable[Number, str], Number, str, Wires): wires to calculate
+                marginal probabilities for. Wires not provided are traced out of one or the system.
+            shot_range (tuple[int]): 2-tuple of integers specifying the range of samples
+                to use. If not specified, all samples are used.
+            bin_size (int): Divides the shot range into bins of size ``bin_size``, and
+                returns the measurement statistic separately over each bin. If not
+                provided, the entire shot range is treated as a single bin.
 
-        return self._prob_array
+        Returns:
+            array[float]: list of the probabilities
+        """
+        try:
+            self._samples = self.generate_samples()
+            result = super().estimate_probability(wires, shot_range, bin_size)
+            self._samples = None
+            return result
+        except TypeError as e:
+            self._samples = None
+            e.args = ("When invoking QubitDevice estimate_probability() method after submitting circuits using any other method than QubitDevice.execute() method, current circuit index should be set via set_current_circuit_index() method pointing to index of the circuit which is targeted out of one or potentially several circuits submitted in a batch.", *e.args)

```

Comment:
Is there something a user can do that makes this come up?

Reply:
Removed, it was a bad idea. The feature I intended to capture here is now handled by CircuitIndexNotSetException which is now also properly tested.